lock-api/src/main/java/com/palantir/lock/watch/ConjureLockRequestMetadataUtils.java

```diff
@@ -0,0 +1,158 @@
+        List<LockDescriptor> keyList = conversionResult.lockList();
+        ConjureLockRequestMetadata conjureMetadata = conversionResult.conjureMetadata();
+        ChecksumType checksumType = ChecksumType.valueOf(conjureMetadata.getChecksumTypeId());
+        KeyListChecksum checksum = KeyListChecksum.of(
+                checksumType, conjureMetadata.getChecksumValue().asNewByteArray());
+        IndexEncodingResult<LockDescriptor, ConjureChangeMetadata> encoded =
+                IndexEncodingResult.<LockDescriptor, ConjureChangeMetadata>builder()
+                        .keyList(keyList)
+                        .indexToValue(conjureMetadata.getIndexToChangeMetadata())
+                        .keyListChecksum(checksum)
+                        .build();
+        Map<LockDescriptor, ChangeMetadata> changeMetadata = IndexEncodingUtils.decode(
+                encoded,
+                conjureChangeMetadata -> conjureChangeMetadata.accept(ChangeMetadataFromConjureVisitor.INSTANCE));
+        // visitUnknown() will return null
+        changeMetadata.values().removeIf(Objects::isNull);
+        return LockRequestMetadata.of(changeMetadata);
+    }
+
+    @Unsafe
+    @Value.Immutable
+    public interface ConjureMetadataConversionResult {
+        List<LockDescriptor> lockList();
+
+        ConjureLockRequestMetadata conjureMetadata();
+
+        static ImmutableConjureMetadataConversionResult.Builder builder() {
+            return ImmutableConjureMetadataConversionResult.builder();
+        }
+    }
+
+    private static final class ChangeMetadataToConjureVisitor implements ChangeMetadata.Visitor<ConjureChangeMetadata> {
+        static final ChangeMetadataToConjureVisitor INSTANCE = new ChangeMetadataToConjureVisitor();
+
+        @Override
+        public ConjureChangeMetadata visit(Unchanged unchanged) {
+            return ConjureChangeMetadata.unchanged(ConjureUnchangedChangeMetadata.of());
+        }
+
+        @Override
+        public ConjureChangeMetadata visit(Updated updated) {
+            return ConjureChangeMetadata.updated(
+                    ConjureUpdatedChangeMetadata.of(Bytes.from(updated.oldValue()), Bytes.from(updated.newValue())));
+        }
+
+        @Override
+        public ConjureChangeMetadata visit(Deleted deleted) {
+            return ConjureChangeMetadata.deleted(ConjureDeletedChangeMetadata.of(Bytes.from(deleted.oldValue())));
+        }
+
+        @Override
+        public ConjureChangeMetadata visit(Created created) {
+            return ConjureChangeMetadata.created(ConjureCreatedChangeMetadata.of(Bytes.from(created.newValue())));
+        }
+    }
+
+    private static final class ChangeMetadataFromConjureVisitor
+            implements ConjureChangeMetadata.Visitor<ChangeMetadata> {
+
+        static final ChangeMetadataFromConjureVisitor INSTANCE = new ChangeMetadataFromConjureVisitor();
```

Comment:
If we need a singleton we should just instead use an ENUM with a single type. Example: https://github.com/palantir/atlasdb/blob/develop/atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/transaction/WitnessToActionVisitor.java

Reply:
oh nice, didn't know that was possible